### PR TITLE
Add SetupDiGetDriverInfoDetail

### PR DESCRIPTION
**Pull request by qmfrederik**

### CODE REVIEW

File: src/SetupApi.Tests/SP_DRVINFO_DETAIL_DATAFacts.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+using static PInvoke.SetupApi;
+using System.Runtime.InteropServices;
+using Xunit;
+
+public class SP_DRVINFO_DETAIL_DATAFacts
+{
+    [Fact]
+    public void Layout_Test()
+    {
+        if (Environment.Is64BitProcess)
```
Comment (qmfrederik) [inline src/SetupApi.Tests/SP_DRVINFO_DETAIL_DATAFacts.cs:16]: These values come from a unit test from our internal codebase. IIRC, they were obtained by calling `sizeof` and `offsetof` the various structs/fields in a C program.

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+            /// <summary>
+            /// The length, in characters, of the CompatIDs list starting at offset <see cref="CompatIDsOffset"/> from the beginning of the <see cref="HardwareID"/> buffer.
+            /// </summary>
+            /// <remarks>
+            /// If <see cref="CompatIDsLength"/> is nonzero, the CompatIDs list contains one or more <see langword="null"/>-terminated strings with an additional
+            /// <see langword="null"/> character at the end of the list.
+            /// If <see cref="CompatIDsLength"/> is zero, the CompatIDs list is empty. In that case, there is no additional <see langword="null"/> character at the end of the list.
+            /// </remarks>
+            public IntPtr CompatIDsLength;
```
Comment (qmfrederik) [inline src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: The documentation says that this field is a `DWORD`, not a pointer.

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            /// </summary>
+            /// <returns>An instance of <see cref="SP_DRVINFO_DETAIL_DATA"/>.</returns>
+            public static SP_DRVINFO_DETAIL_DATA Create()
+            {
+                var size = Marshal.SizeOf(typeof(SP_DRVINFO_DETAIL_DATA));
+
+                if (!Environment.Is64BitProcess)
+                {
+                    // On 32-bit Windows
+                    size -= 6;
```
Comment (qmfrederik) [inline src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: IIRC, we got this struct to work "without tricks" on 32-bit Windows by setting `Pack = 1` in the `StructLayout` attribute, and `HardwareID[1]`, and `Pack = 4` and `HardwareId[4]` for 64-bit Windows.

**Comment by qmfrederik** [comment]
@AArnott The `SP_DRVINFO_DETAIL_DATA` struct is a bit of a mystery to me. The code in this PR makes the tests pass on 32-bit and 64-bit Windows; but I'm not quite sure why the magic which is being done works/is required.

As a sidenote, I think the unit tests in this repo only run in a 32-bit process, correct?

**Comment by AArnott** [comment]
> I think the unit tests in this repo only run in a 32-bit process, correct?

Yes, but I should fix this: #486

File: src/SetupApi.Tests/SP_DRVINFO_DETAIL_DATAFacts.cs (hunk 1 above)
Comment (AArnott) [reply src/SetupApi.Tests/SP_DRVINFO_DETAIL_DATAFacts.cs:16]: That's a decent code comment to add (explaining that these values came from a C program) so we know we can trust them.

File: src/SetupApi.Tests/SetupApiFacts.cs [hunk 4]
```diff
@@ -159,4 +159,64 @@ public void SetupDiEnumDriverInfoListTest()
...
+            var loopbackDrivers =
+                driverInfos
+                .Where(d => d.DescriptionString.IndexOf("loopback", StringComparison.OrdinalIgnoreCase) >= 0).ToArray();
+
+            var loopbackDriver = Assert.Single(loopbackDrivers);
+
+            const int bufferSize = 0x1000;
+
+            IntPtr drvDetailDataBuffer = Marshal.AllocHGlobal(bufferSize);
+            Marshal.StructureToPtr(SP_DRVINFO_DETAIL_DATA.Create(), drvDetailDataBuffer, false);
```
Comment (AArnott) [inline src/SetupApi.Tests/SetupApiFacts.cs]: Tip: code actually gets simpler (and you don't have to worry about your missing try/finally to avoid memory leaks on exceptions) by using ordinary `byte[]` and pointers. There's no reason to call `AllocHGlobal` and `FreeHGlobal` to get this memory.

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs (hunk 2 above)
Comment (AArnott) [reply src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: The doc agrees with the header file:
image

I'm very surprised to see `IntPtr` here, and to see your test that is backed by a C program validate that the offset for `Reserved` is correct. Maybe it's just an alignment thing and your `Pack` value is off?
In fact it has to be an alignment issue of some sort because in a C program, `sizeof(CompatIDsLength)` returns 4 in both x86 and x64. So we should *not* claim it's IntPtr size here. Let's figure out the alignment issue another way.

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            /// </summary>
+            /// <remarks>
+            /// Each ID in the list is a <see langword="null"/>-terminated string.
+            /// If the hardware ID exists (that is, if <see cref="CompatIDsOffset"/> is greater than one), this single <see langword="null"/>-terminated
+            /// string is found at the beginning of the buffer.
+            /// If the CompatIDs list is not empty (that is, if <see cref="CompatIDsLength"/> is not zero), the CompatIDs list starts at offset
+            /// <see cref="CompatIDsOffset"/> from the beginning of this buffer, and is terminated with an additional <see langword="null"/>
+            /// character at the end of the list.
+            /// </remarks>
+            public fixed char HardwareID[4];
```
Comment (AArnott) [inline src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: Why is this `4` instead of the customary `1` for `ANYSIZE_ARRAY`?

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            /// This must be the basic DDInstall section name, such as InstallSec, without any OS/architecture-specific extensions.
+            /// </summary>
+            public string SectionNameString
+            {
+                get
+                {
+                    fixed (char* sectionName = this.SectionName)
+                    {
+                        return new string(sectionName);
+                    }
```
Comment (AArnott) [inline src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: I'm not sure why at one point C# told me I had to used `fixed` around these, but it isn't necessary. On your CfgMgr32 PR [I removed it](https://github.com/dotnet/pinvoke/pull/473/commits/0f4a2b862fed53e68ec751f5e2c19df1607e9e0d) and `new string(this.SectionName)` should just work. So given how easy it is, I favor removing these helper properties, particularly since they allocate on every call which generally isn't something that callers expect.

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs (hunk 3 above)
Comment (AArnott) [reply src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: Ya, not a fan of these tricks, particularly when I don't understand why they're necessary since I don't know when this might change to something that breaks (e.g. Windows on ARM)?

File: src/SetupApi/SetupApi.csproj [hunk 7]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="MSBuild.Sdk.Extras">
   <PropertyGroup>
-    <TargetFrameworks>$(PlatformAndPortableFrameworks)</TargetFrameworks>
+    <TargetFrameworks>net45;netstandard2.0</TargetFrameworks>
```
Comment (AArnott) [inline src/SetupApi/SetupApi.csproj]: Why is this necessary, particularly separately from #476?

**Review by AArnott — CHANGES_REQUESTED** [review]

**Comment by qmfrederik** [comment]
For the sake of completeness, here's the source code for that test program, and the output when ran as a 32-bit or 64-bit program:

```c
#include <iostream>
#include <Windows.h>
#include "setupapi.h"

int main()
{
    printf("Bitness:                                 0x%p\n", sizeof(void*));
    printf("\n");
    printf("SP_DRVINFO_DETAIL_DATA size:             0x%p\n", sizeof(SP_DRVINFO_DETAIL_DATA));
    printf("\n");
    printf("SP_DRVINFO_DETAIL_DATA.cbSize          @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, cbSize));
    printf("SP_DRVINFO_DETAIL_DATA.InfDate         @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, InfDate));
    printf("SP_DRVINFO_DETAIL_DATA.CompatIDsOffset @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, CompatIDsOffset));
    printf("SP_DRVINFO_DETAIL_DATA.CompatIDsLength @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, CompatIDsLength));
    printf("SP_DRVINFO_DETAIL_DATA.Reserved        @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, Reserved));
    printf("SP_DRVINFO_DETAIL_DATA.SectionName     @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, SectionName));
    printf("SP_DRVINFO_DETAIL_DATA.InfFileName     @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, InfFileName));
    printf("SP_DRVINFO_DETAIL_DATA.DrvDescription  @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, DrvDescription));
    printf("SP_DRVINFO_DETAIL_DATA.HardwareID      @ 0x%p\n", offsetof(SP_DRVINFO_DETAIL_DATA, HardwareID));
}
```

```
Bitness:                                 0x00000004

SP_DRVINFO_DETAIL_DATA size:             0x00000622

SP_DRVINFO_DETAIL_DATA.cbSize          @ 0x00000000
SP_DRVINFO_DETAIL_DATA.InfDate         @ 0x00000004
SP_DRVINFO_DETAIL_DATA.CompatIDsOffset @ 0x0000000C
SP_DRVINFO_DETAIL_DATA.CompatIDsLength @ 0x00000010
SP_DRVINFO_DETAIL_DATA.Reserved        @ 0x00000014
SP_DRVINFO_DETAIL_DATA.SectionName     @ 0x00000018
SP_DRVINFO_DETAIL_DATA.InfFileName     @ 0x00000218
SP_DRVINFO_DETAIL_DATA.DrvDescription  @ 0x00000420
SP_DRVINFO_DETAIL_DATA.HardwareID      @ 0x00000620
```

```
Bitness:                                 0x0000000000000008

SP_DRVINFO_DETAIL_DATA size:             0x0000000000000630

SP_DRVINFO_DETAIL_DATA.cbSize          @ 0x0000000000000000
SP_DRVINFO_DETAIL_DATA.InfDate         @ 0x0000000000000004
SP_DRVINFO_DETAIL_DATA.CompatIDsOffset @ 0x000000000000000C
SP_DRVINFO_DETAIL_DATA.CompatIDsLength @ 0x0000000000000010
SP_DRVINFO_DETAIL_DATA.Reserved        @ 0x0000000000000018
SP_DRVINFO_DETAIL_DATA.SectionName     @ 0x0000000000000020
SP_DRVINFO_DETAIL_DATA.InfFileName     @ 0x0000000000000220
SP_DRVINFO_DETAIL_DATA.DrvDescription  @ 0x0000000000000428
SP_DRVINFO_DETAIL_DATA.HardwareID      @ 0x0000000000000628
```

It doesn't answer the _why_ but serves as a confirmation that the values in that unit test are correct.

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs (hunk 2 above)
Comment (qmfrederik) [reply src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: Looking at this declaration once more, I believe this is because [pointers (i.e. `ULONG_PTR Reserved` in this case) are aligned to 32-bit (or 4 byte) boundaries on x86 and 64-bit (or 8 byte) boundaries on x64](https://docs.microsoft.com/en-us/windows/win32/midl/c-compiler-packing-issues#natural-alignment).

`cbSize`, `CompatIDsOffset` and `CompatIDsLength` are 4 bytes wide; `FILETIME` 8 bytes, so that's a total of 4 + 8 + 4 + 4 = 20 bytes, which is a multiple of 4 but not of 8.

Comment (qmfrederik) [reply src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: OK - so the solution simply is to use `Pack = 0` (or omit the `Pack` value alltogether).

File: src/SetupApi/SetupApi.csproj (hunk 7 above)
Comment (qmfrederik) [reply src/SetupApi/SetupApi.csproj]: I wanted to know whether the code was running in 64-bit mode or not, so I used `Environment.Is64BitProcess`. That's not available on `netstandard1.x` so instead of using `IntPtr.Size`, I just disabled `netstandard1.x`. support. It was late 😄 .

File: src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs (hunk 2 above)
Comment (vatsan-madhavan) [reply src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs]: For structs that require specific alignment, Win32 headers tend to do one of the two things - 

- Use `__declspec(align(alignment_size))` or `DECLSPEC_ALIGN` macro. `DirectXMath.h` has several types that leverages this, like `XMMATRIX`. 
- Use padding fields. Quick scan of `setupapi.h` shows that `_SP_SELECTDEVICE_PARAMS_A.Reserved` is one such field. 

Outside of these patterns, its rare to see special alignment considerations pop-up that are purely *inferred* contextually but aren't encoded in the struct-definition in any other manner. It can happen (there are bugs, failure to document, implicit assumptions etc.), but its rare. 

Aside: Sometimes you'll see (C++) `sizeof` return a value larger than the sum of individual member `sizeof` values - which can happen due to default alignment rules associated with individual types. For e.g, `{DWORD, CHAR, DWORD}` would likely return a `sizeof` `12` instead of `9 (4 + 1 + 4)` (`Marshal.SizeOf` usually returns `12` correctly in these situations AFAICT).

**Review by AArnott — APPROVED** [review]
